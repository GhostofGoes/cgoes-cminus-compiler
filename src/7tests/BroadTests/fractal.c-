// Author: Jeremy Scott
// Spring 2006, CS 445
// Code to generate a ASCII Mandelbrot & Julia fractals using the tiny machine

// DEFAULT_TOP    -1.25
// DEFAULT_LEFT   -2.0
// DEFAULT_HEIGHT  2.5
// DEFAULT_WIDTH   2.8
// Multiply all of these by 256 to get fixed-point values...

// Since the tiny machine doesn't support floating point, we have to do everything in fixed-point
// i.e. 
// 1 == 1 * 256 == 0x100
// So for fixed-point multiplication, 1 * 1 == (1 * 256) * (1 * 256) == (1 * 65536)
// Therefore, after any multipy, we have to divide by 256 to restore the decimal point to the right spot
//
// To better see the sets you might try doing this UNIX command:
//
// tm fractal  |  tr '0123456789FT' ' .,;-+oOGM@//'
//

int TOP;
int LEFT;
int HEIGHT;
int WIDTH;
int XStep;
int YStep;
int XMax, YMax;
int iter;

// max length = 4 * 256 == 10024
  Mandelbrot(int JuliaVals[])
{
    while (TOP < YMax)
    {
        LEFT = -512;
        while (LEFT < XMax)
        {
            int Len;
            int Bi, A, NextA, NextBi;
            A = LEFT;
            Bi = TOP;
            NextA = (A * A) / 256;
            NextBi = (Bi * Bi) / 256;
            Len = NextA + NextBi;
            iter = 0;
            while (Len <= 1024  &  iter < 11)
            {
                Bi = 2 * (A * Bi) / 256 + TOP;
                A = NextA - NextBi + LEFT;
                NextA = (A * A) / 256;
                NextBi = (Bi * Bi) / 256;
                Len = NextA + NextBi;
                iter++;
            }
            if (iter < 10)
                output(iter);
            else
                outputb(iter == 11);
            if (iter == 11)
            {
                JuliaVals[0] = LEFT;
                JuliaVals[1] = TOP;
            }
            LEFT += XStep;
        }
        outnl();
        TOP += YStep;
    }
}

  Julia(int CA; int CBi)
{
    while (TOP < YMax)
    {
        LEFT = -512;
        while (LEFT < XMax)
        {
            int Len;
            int Bi, A, NextA, NextBi;
            A = LEFT;
            Bi = TOP;
            NextA = (A * A) / 256;
            NextBi = (Bi * Bi) / 256;
            Len = NextA + NextBi;
            iter = 0;
            while (Len <= 1024  &  iter < 11)
            {
                Bi = 2 * (A * Bi) / 256 + CBi;
                A = NextA - NextBi + CA;
                NextA = (A * A) / 256;
                NextBi = (Bi * Bi) / 256;
                Len = NextA + NextBi;
                iter++;
            }
            if (iter < 10)
                output(iter);
            else
                outputb(iter == 11);
            LEFT += XStep;
        }
        outnl();
        TOP += YStep;
    }
}

main()
{
    int JuliaVals[2];
    TOP = -320;
    LEFT = -512;
    HEIGHT = 640;
    WIDTH = 716;
    XStep = WIDTH / 40;
    YStep = HEIGHT / 36;
    XMax = LEFT + WIDTH;
    YMax = TOP + HEIGHT;
    Mandelbrot(JuliaVals);

    outnl();
	
    TOP = -320;
    LEFT = -512;
    HEIGHT = 640;
    WIDTH = 716;
    XStep = WIDTH / 40;
    YStep = HEIGHT / 36;
    XMax = LEFT + WIDTH;
    YMax = TOP + HEIGHT;
    Julia(JuliaVals[0], JuliaVals[1]);
}
