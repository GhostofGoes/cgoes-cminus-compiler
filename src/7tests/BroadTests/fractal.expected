Number of warnings: 0
Number of errors: 0
------------------------------------

Memory Configuration: Data Addresses: 0-9999  Instruction Addresses: 0-9999
Loading file: fractal.tm
0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 1 1 1 1 1 1 1 1 1 1 1 1 
0 0 0 0 0 0 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 1 1 1 1 1 1 1 1 1 1 
0 0 0 0 0 0 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 F 4 4 3 3 2 2 2 1 1 1 1 1 1 1 1 
0 0 0 0 0 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 4 4 6 9 5 4 3 3 3 2 2 2 1 1 1 1 1 1 
0 0 0 0 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 4 4 6 T T 7 4 3 3 3 2 2 2 1 1 1 1 1 
0 0 0 0 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 4 4 6 7 T T 6 5 4 3 3 3 2 2 2 1 1 1 1 
0 0 0 1 1 1 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 3 4 4 5 T T T T T 6 4 4 3 3 3 2 2 2 1 1 1 
0 0 0 1 1 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 4 4 5 5 6 F T T T T 6 5 4 4 4 3 2 2 2 2 1 1 
0 0 1 1 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 4 5 6 6 6 8 F T T T T 8 6 5 5 5 4 3 2 2 2 2 1 
0 0 1 2 2 2 2 2 2 2 2 2 2 2 3 3 3 4 4 4 5 T T 9 T T T T T T T T T 8 7 8 6 4 3 2 2 2 2 
0 0 1 2 2 2 2 2 2 2 2 2 2 3 3 4 4 4 4 5 6 9 T T T T T T T T T T T T T T T 4 3 2 2 2 2 
0 0 2 2 2 2 2 2 2 2 2 3 3 4 4 4 4 4 5 5 7 F T T T T T T T T T T T T T T 7 4 3 2 2 2 2 
0 1 2 2 2 2 2 2 2 3 3 5 5 5 4 4 5 5 6 7 T T T T T T T T T T T T T T T T 7 5 4 3 2 2 2 
0 2 2 2 2 2 3 3 3 4 5 F 7 6 7 7 6 6 7 8 T T T T T T T T T T T T T T T T T T 4 3 2 2 2 
0 2 2 2 3 3 3 3 4 4 5 9 T F F T T 8 8 T T T T T T T T T T T T T T T T T T 8 4 3 2 2 2 
0 2 3 3 3 3 3 4 4 4 6 8 T T T T T T F T T T T T T T T T T T T T T T T T T 6 4 3 3 2 2 
0 3 3 3 3 3 4 4 4 6 7 T T T T T T T T T T T T T T T T T T T T T T T T T T 6 4 3 3 2 2 
0 3 3 3 3 4 5 5 6 T T T T T T T T T T T T T T T T T T T T T T T T T T T F 5 4 3 3 2 2 
1 4 5 6 6 6 6 7 9 T T T T T T T T T T T T T T T T T T T T T T T T T T T 6 4 4 3 3 2 2 
2 6 7 9 F F F F T T T T T T T T T T T T T T T T T T T T T T T T T T T 9 6 4 4 3 3 2 2 
0 4 4 3 5 5 5 6 7 F T T T T T T T T T T T T T T T T T T T T T T T T T T 7 5 4 3 3 2 2 
0 3 3 3 3 4 4 5 6 T 8 T T T T T T T T T T T T T T T T T T T T T T T T T T 5 4 3 3 2 2 
0 2 3 3 3 3 4 4 4 5 6 9 T T T T T T T T T T T T T T T T T T T T T T T T T 6 4 3 3 2 2 
0 2 2 3 3 3 3 4 4 4 6 7 T T T T T T 9 T T T T T T T T T T T T T T T T T T 6 4 3 2 2 2 
0 2 2 2 3 3 3 3 4 4 5 T 8 8 9 F 8 7 7 F T T T T T T T T T T T T T T T T T T 4 3 2 2 2 
0 1 2 2 2 2 2 3 3 3 5 T 6 5 6 6 6 6 6 7 T T T T T T T T T T T T T T T T 8 T 4 3 2 2 2 
0 1 2 2 2 2 2 2 2 2 3 4 4 4 4 4 4 5 5 6 T T T T T T T T T T T T T T T T 6 5 4 3 2 2 2 
0 0 2 2 2 2 2 2 2 2 2 2 3 4 4 4 4 4 5 5 7 9 T T T T T T T T T T T T T T 8 4 3 2 2 2 2 
0 0 1 2 2 2 2 2 2 2 2 2 2 3 3 3 4 4 4 5 6 T T T T T T T T T T T T F T T T 4 3 2 2 2 2 
0 0 1 1 2 2 2 2 2 2 2 2 2 2 3 3 3 3 4 4 5 T T 8 9 T T T T T T T T 7 6 7 6 3 2 2 2 2 2 
0 0 0 1 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 4 4 5 5 6 7 T T T T T 7 6 5 5 4 3 3 2 2 2 2 1 
0 0 0 1 1 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 4 4 5 6 9 T T T T 6 4 4 4 3 3 2 2 2 2 1 1 
0 0 0 0 1 1 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 4 4 5 7 F T T 8 5 4 4 3 3 3 2 2 2 1 1 1 
0 0 0 0 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 4 4 5 7 F F 6 4 3 3 3 3 2 2 2 1 1 1 1 
0 0 0 0 0 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 4 4 6 T 7 T 3 3 3 3 2 2 2 1 1 1 1 1 
0 0 0 0 0 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 4 7 6 5 4 3 3 2 2 2 1 1 1 1 1 1 1 
0 0 0 0 0 0 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 4 4 3 3 2 2 2 2 1 1 1 1 1 1 1 1 
0 0 0 0 0 0 0 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 1 1 1 1 1 1 1 1 1 1 

0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 
0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 3 3 3 3 3 3 3 3 2 2 2 2 
0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 3 3 4 T 5 4 3 3 3 3 3 3 3 3 
0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 3 3 4 6 T 5 4 4 3 3 3 3 3 3 3 
0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 3 3 3 5 8 9 6 4 4 4 3 3 3 3 3 4 
0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 3 3 3 4 7 T 7 6 5 4 4 4 4 4 4 5 
0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 3 3 3 4 4 5 F T 9 8 5 5 5 5 5 6 T 
0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 3 3 3 4 4 5 6 8 T 7 6 6 5 5 6 9 T 
0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 3 3 3 4 4 5 6 7 T F 7 6 6 7 9 T 8 
0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 3 3 4 4 4 5 6 7 F T 9 9 T 9 T F 6 
0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 3 3 3 4 4 5 5 7 9 T T T T 9 T 8 6 5 
0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 3 3 4 4 5 5 6 8 T 9 8 8 7 7 8 9 5 4 
0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 3 3 4 6 6 6 7 9 T 7 7 6 6 6 6 5 4 4 
0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3 6 F 8 8 8 T F 7 6 5 5 5 5 4 4 3 
0 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3 4 5 7 F T T F T 6 5 5 5 4 4 4 3 3 
0 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 3 3 4 5 6 8 T F 7 7 5 5 5 4 4 4 3 3 3 
0 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3 3 4 5 6 7 T 9 7 6 5 5 4 4 4 3 3 3 3 
0 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 3 3 4 4 5 6 T T 9 6 6 5 4 4 4 4 3 3 3 2 
1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 3 3 4 4 4 5 6 8 T T 6 5 4 4 4 4 3 3 3 2 2 
1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3 3 4 4 4 5 6 9 T T 7 5 4 4 4 3 3 3 2 2 2 
0 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3 3 4 4 4 4 5 6 F T T 7 5 4 4 4 3 3 2 2 2 2 
0 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 3 3 3 3 4 4 4 4 5 6 6 8 T 9 7 5 4 4 3 3 2 2 2 2 2 
0 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 3 3 3 3 4 4 4 4 5 5 6 7 8 T 7 6 5 4 4 3 3 2 2 2 2 2 
0 1 1 1 1 1 1 1 1 2 2 2 2 2 2 3 3 3 3 3 3 4 4 4 5 5 5 6 8 F T 8 6 5 4 3 3 2 2 2 2 2 2 
0 1 1 1 1 1 1 1 1 2 2 2 2 2 3 3 3 3 3 3 4 4 5 5 5 5 6 8 T T F T T 6 4 3 3 2 2 2 2 2 2 
0 1 1 1 1 1 1 1 2 2 2 2 3 3 3 3 3 3 3 4 4 5 5 5 5 6 7 F T 8 7 6 7 6 3 3 2 2 2 2 2 2 2 
0 1 1 1 1 1 1 1 2 2 2 3 3 3 3 3 3 3 4 4 6 7 6 6 6 7 8 T F 7 6 5 5 4 3 3 2 2 2 2 2 2 2 
0 0 1 1 1 1 1 1 2 2 2 3 3 3 3 4 4 4 5 5 7 T 8 8 T 8 9 T F 6 5 5 4 4 3 3 2 2 2 2 2 2 2 
0 0 1 1 1 1 1 1 2 2 3 4 4 4 4 4 5 5 5 7 8 T T T T T T 8 7 5 5 4 4 3 3 3 2 2 2 2 2 2 2 
0 0 1 1 1 1 1 1 2 2 4 7 6 5 5 6 6 6 7 F T 8 9 8 8 F F 7 6 5 4 4 4 3 3 3 2 2 2 2 2 2 1 
0 0 0 1 1 1 1 1 2 2 4 9 T T T 7 7 8 8 T 8 7 6 6 6 7 T 8 6 5 4 4 3 3 3 2 2 2 2 2 2 1 1 
0 0 0 1 1 1 1 1 1 2 3 5 5 6 7 T T F T 7 6 5 5 5 6 7 T F 7 5 4 4 3 3 3 2 2 2 2 2 1 1 1 
0 0 0 0 1 1 1 1 1 2 3 3 4 4 6 T 7 6 9 6 5 5 5 5 5 5 7 9 T 8 4 4 3 3 3 2 2 2 2 1 1 1 1 
0 0 0 0 1 1 1 1 1 2 2 3 3 4 5 F 5 5 4 4 4 3 4 4 4 4 5 6 9 7 5 3 3 3 2 2 2 2 1 1 1 1 1 
0 0 0 0 0 1 1 1 1 1 2 2 3 3 7 6 4 4 3 3 3 3 3 3 4 4 4 5 8 T T 3 3 3 2 2 2 1 1 1 1 1 1 
0 0 0 0 0 1 1 1 1 1 2 2 2 2 3 3 3 3 3 3 3 3 3 3 3 3 4 5 7 5 4 3 3 2 2 1 1 1 1 1 1 1 1 
0 0 0 0 0 0 1 1 1 1 1 2 2 2 2 2 2 2 2 2 3 3 3 3 3 3 3 4 T 4 3 2 2 2 1 1 1 1 1 1 1 1 1 
0 0 0 0 0 0 0 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 2 2 2 2 1 1 1 1 1 1 1 1 1 1 
Status: Halted
Bye.
